File: server/lib/capability-document.mjs
/**
 * What this build can actually do — one document, assembled from the code.
 *
 * WHY this file exists: "what can move money here, and what is simulated?" is a
 * fair question with an embarrassing failure mode — the answer was a transfer
 * prompt. The honest answer is a list of the app's real capabilities, and a
 * hand-written list goes stale the first time a verb changes. So the document
 * is assembled from the catalogues that already drive behaviour:
 *
 *   · `verbs.mjs`           — every typed verb, by class (observe / prepare /
 *                             commit). The three middle sections are exactly
 *                             that catalogue, filtered by class; a verb cannot
 *                             exist without appearing here.
 *   · `capabilities.mjs`    — the engines that run as research tasks, with
 *                             their real connector state (none connected).
 *   · `corridor-support.mjs` and `rate-guard.mjs` — the refusals, with the
 *                             lines the app actually says.
 *
 * The same five sections are mirrored in the app (`CapabilityDocument.swift`),
 * so a question asked with the proxy down gets the same answer.
 */

import { catalogue } from "./verbs.mjs";
import { capabilityStates } from "./capabilities.mjs";
import { quotedCurrenciesPhrase, quotedPairsPhrase, unquotedCodes } from "./corridor-support.mjs";
import { RATE_BOOKING_REFUSAL } from "./rate-guard.mjs";
import { REFUSAL_LINE } from "./injection-guard.mjs";

export const CAPABILITY_DOCUMENT_VERSION = 1;

/** The refusal line for investment, tax and legal advice. One wording. */
export const ADVICE_REFUSAL =
  "That is investment advice and I do not give it — nobody should, without knowing your whole position. " +
  "I can show you what you hold, what a fee costs you, and the public facts about anything you name.";

/** The line used when a model-written reply may not be trusted. */
export const UNGROUNDED_REFUSAL = "Let me not put a number on that without checking your account first.";

/** The five sections, in the order a person would ask about them. */
export const CAPABILITY_SECTIONS = Object.freeze([
  Object.freeze({ id: "instant", label: "Answered here and now, from your own records" }),
  Object.freeze({ id: "prepares", label: "Prepared, with nothing moved" }),
  Object.freeze({ id: "approval", label: "Moves only after one approval" }),
  Object.freeze({ id: "research", label: "Runs as research, on the Mac runtime, preparation only" }),
  Object.freeze({ id: "refuses", label: "Refused, deterministically" }),
]);

/**
 * A human name per verb. Keyed by the verb ids in `verbs.mjs`; the test asserts
 * this table covers every id in the catalogue, so a new verb cannot be added
 * without either naming it or failing.
 */
const VERB_LABELS = Object.freeze({
  "balance.read": "your balances",
  "budget.read": "this week's budget and the reserve",
  "card.read": "the card and its controls",
  "receive.read": "your receiving details",
  "question.answer": "a short explanation or a follow-up",
  "task.read": "the state of a running task",
  "desk.report": "a desk report — fees, offers, claims",
  "transfer.prepare": "a transfer, ready for your confirmation",
  "budget.prepare": "a change to the weekly budget",
  "task.create": "a research or comparison task",
  "watch.create": "a standing watch on something you name",
  "requirements.prepare": "the requirements checklist when no provider is connected",
  "transfer.commit": "sending money to a saved recipient",
  "fx.swap": "a currency swap at a rate we quoted",
  "card.freeze": "freezing or unfreezing the card",
  "card.block": "blocking a merchant on the card",
  "subscription.cancel": "cancelling a subscription",
});

/** The refusals, with the words the app actually uses. */
function refusals() {
  return [
    {
      id: "advice",
      label: "investment, tax and legal advice",
      line: ADVICE_REFUSAL,
    },
    {
      id: "rate_not_quoted",
      label: "a rate we did not quote",
      line: RATE_BOOKING_REFUSAL,
    },
    {
      id: "unpriced_corridor",
      label: `a corridor this build does not price (${unquotedCodes().slice(0, 6).join(", ")}, …)`,
      line: `Any pair among ${quotedCurrenciesPhrase()}. I will not invent a rate outside them.`,
    },
    {
      id: "rules_override",
      label: "instructions that try to change my rules, limits or approvals",
      line: REFUSAL_LINE,
    },
    {
      id: "invented_figure",
      label: "an invented figure in a model-written line",
      line: UNGROUNDED_REFUSAL,
    },
  ];
}

/**
 * The document, built from the catalogues. Nothing here is transcribed by hand
 * except the labels, which the tests hold to the catalogue's own ids.
 */
export function capabilityDocument() {
  const verbs = catalogue().verbs;
  const byClass = (verbClass) => verbs.filter((verb) => verb.class === verbClass);
  const item = (verb) => ({
    id: verb.id,
    label: VERB_LABELS[verb.id] ?? verb.id,
    detail: verb.route,
    requires: [...verb.requires],
  });

  const engines = capabilityStates();

  return {
    version: CAPABILITY_DOCUMENT_VERSION,
    sections: [
      {
        id: "instant",
        label: CAPABILITY_SECTIONS[0].label,
        items: byClass("observe").map(item),
      },
      {
        id: "prepares",
        label: CAPABILITY_SECTIONS[1].label,
        items: byClass("prepare").map(item),
      },
      {
        id: "approval",
        label: CAPABILITY_SECTIONS[2].label,
        items: byClass("commit").map(item),
      },
      {
        id: "research",
        label: CAPABILITY_SECTIONS[3].label,
        items: engines.map((engine) => ({
          id: `task.${engine.id}`,
          label: engine.label,
          detail: `research and preparation only; connector ${engine.connector.status}`,
        })),
      },
      {
        id: "refuses",
        label: CAPABILITY_SECTIONS[4].label,
        items: refusals(),
      },
    ],
    corridors: {
      quoted: quotedPairsPhrase(),
      currencies: quotedCurrenciesPhrase(),
      unquotedExamples: unquotedCodes(),
    },
    financialMode: "SIMULATED",
  };
}

/** The document's items for one section, or an empty list. */
export function sectionItems(document, id) {
  return document.sections.find((section) => section.id === id)?.items ?? [];
}

/**
 * A deterministic read for the obvious shapes of "tell me about this build":
 * what it can do, what moves money, what needs approval, what is simulated,
 * what it refuses. The typed question in the batch catches paraphrases; this
 * narrow shape answers the plain phrasings even when the provider is down.
 */
const ABOUT_THE_BUILD =
  /\b(?:what can (?:you|this|the app|this build|the product|it|mira)|what (?:does|do) (?:this|you|the app|mira|the build|it)(?:\s+\w+){0,2}\s+(?:do|offer|support|handle)|what (?:can't|can’t|cannot|don't|do not|doesn't|does not) (?:you|this|the app|mira|it|this build)|what (?:actually )?moves? money|move money (?:in|here|in this build)|what(?:'s| is) simulated|what needs (?:my|your|one|an?) approval|what do you do|capabilit(?:y|ies)|what are you (?:able|allowed) to do|how does (?:mira|this|the app) work|what does the user see)\b/i;

export function asksAboutThisBuild(text) {
  return ABOUT_THE_BUILD.test(String(text || ""));
}

/**
 * The one-paragraph answer to "what can this build do?". It is prose over the
 * document, never a second list: the items it names are the document's own.
 */
export function renderCapabilityAnswer(document = capabilityDocument()) {
  const names = (id) =>
    sectionItems(document, id)
      .map((entry) => entry.label)
      .join(", ");
  const research = sectionItems(document, "research").length;
  return [
    `This build is simulated money, and it is honest about the line. Instantly, from your own records: ${names("instant")}.`,
    `Prepared, with nothing moved: ${names("prepares")}.`,
    `One approval, then it moves: ${names("approval")}.`,
    `${research} engines run as research on the Mac runtime, and every one of them is preparation only — no vendor is connected.`,
    `I refuse ${sectionItems(document, "refuses")
      .map((entry) => entry.label)
      .join(", ")}.`,
  ].join(" ");
}

File: MiraApp/Mira/Domain/CapabilityDocument.swift
import Foundation

// MARK: - What this build can actually do
//
// The app mirror of `server/lib/capability-document.mjs`, and the answer to
// "what can move money in this build today, and what is simulated?". It is
// assembled from the catalogues this code already runs on — the typed actions
// (`AgentAction.Kind`), the money-desk engines, the commit paths and the
// refusals — so a new action or engine cannot be added without a test failing
// until it is placed in a section.
//
// Five sections, in the order a person asks about them: what is instant, what
// is prepared, what moves after one approval, what runs as research, and what
// is refused. Money is simulated in every one of them.

/// One line of the document: an id the code uses, and the words a person reads.
struct CapabilityItem: Equatable, Sendable {
  let id: String
  let label: String
}

enum CapabilitySectionID: String, CaseIterable, Sendable {
  case instant
  case prepares
  case approval
  case research
  case refuses

  var label: String {
    switch self {
    case .instant: return "Answered here and now, from your own records"
    case .prepares: return "Prepared, with nothing moved"
    case .approval: return "Moves only after one approval"
    case .research: return "Runs as research, on the Mac runtime, preparation only"
    case .refuses: return "Refused, deterministically"
    }
  }
}

/// Which half of the product an action belongs to. The names match the server's
/// verb classes (`verbs.mjs`): observe reads, prepare composes, commit moves.
enum CapabilityClass: String, Sendable {
  case observe
  case prepare
  case commit
}

/// The twelve money-desk engines, each with the flow id the session uses. This
/// is the desk's own catalogue: `flow` is the string passed to `appendMira`.
enum MoneyDeskEngine: String, CaseIterable, Sendable {
  case zombies
  case fees
  case idleCash = "idle-cash"
  case negotiation
  case income
  case split
  case credit
  case flaggedCharge = "flagged-charge"
  case priceClaim = "price-claim"
  case claims
  case agentBudget = "agent-budget"
  case goals

  var label: String {
    switch self {
    case .zombies: return "the charges nobody uses"
    case .fees: return "a fee radar over three months"
    case .idleCash: return "what is safe to put away"
    case .negotiation: return "a renewal ask, drafted"
    case .income: return "income carved into tax, buffer and spendable"
    case .split: return "a split, laid out for the people in it"
    case .credit: return "what keeps credit utilisation under the line"
    case .flaggedCharge: return "a charge you do not recognise, reviewed"
    case .priceClaim: return "a price-drop claim, prepared"
    case .claims: return "a compensation claim, prepared"
    case .agentBudget: return "what an agent budget still allows"
    case .goals: return "a warning when a purchase touches a goal"
    }
  }

  var capabilityClass: CapabilityClass {
    switch self {
    case .negotiation, .income, .split, .priceClaim, .claims: return .prepare
    case .zombies, .fees, .idleCash, .credit, .flaggedCharge, .agentBudget, .goals: return .observe
    }
  }
}

/// The commit paths this app owns. The first five ids are the server's commit
/// verbs; `checkout.commit` is the order flow the app runs itself.
enum CommitVerb: String, CaseIterable, Sendable {
  case transfer = "transfer.commit"
  case swap = "fx.swap"
  case cardFreeze = "card.freeze"
  case cardBlock = "card.block"
  case subscriptionCancel = "subscription.cancel"
  case checkout = "checkout.commit"

  var label: String {
    switch self {
    case .transfer: return "sending money to a saved recipient"
    case .swap: return "a currency swap at a rate we quoted"
    case .cardFreeze: return "freezing or unfreezing the card"
    case .cardBlock: return "blocking a merchant on the card"
    case .subscriptionCancel: return "cancelling a subscription"
    case .checkout: return "placing an order from a checkout"
    }
  }
}

/// The research engines the Mac runtime runs. Mirrors the ids and labels in
/// `server/lib/capabilities.mjs`; both are research and preparation only, and
/// no vendor is connected.
enum ResearchEngine: String, CaseIterable, Sendable {
  case restaurant
  case shopping
  case travel
  case auction
  case invest
  case watch
  case research
  case admin

  var label: String {
    switch self {
    case .restaurant: return "restaurant discovery and reservation preparation"
    case .shopping: return "product research and purchase preparation"
    case .travel: return "travel research and booking preparation"
    case .auction: return "auction search and bidding preparation"
    case .invest: return "investment research and order preparation"
    case .watch: return "a standing watch and price alerts"
    case .research: return "general research and comparison"
    case .admin: return "personal admin and planning"
    }
  }
}

/// The refusals, with the exact words the app uses. The lines come from the
/// guards themselves — one wording each, so the document cannot drift from
/// what is actually said.
enum CapabilityRefusal: String, CaseIterable, Sendable {
  case advice
  case rateNotQuoted = "rate_not_quoted"
  case unpricedCorridor = "unpriced_corridor"
  case rulesOverride = "rules_override"
  case inventedFigure = "invented_figure"

  var label: String {
    switch self {
    case .advice: return "investment, tax and legal advice"
    case .rateNotQuoted: return "a rate we did not quote"
    case .unpricedCorridor: return "a corridor this build does not price"
    case .rulesOverride: return "instructions that try to change my rules, limits or approvals"
    case .inventedFigure: return "an invented figure in a model-written line"
    }
  }

  var line: String {
    switch self {
    case .advice: return StandaloneAgent.adviceRefusal
    case .rateNotQuoted: return StandaloneAgent.rateBookingRefusal
    case .unpricedCorridor: return "Any pair among \(CorridorSupport.quotedCurrenciesPhrase). "
        + "I will not invent a rate outside them."
    case .rulesOverride: return MiraRouteClient.refusalLine
    case .inventedFigure: return "Let me not put a number on that without checking your account first."
    }
  }
}

enum CapabilityDocument {
  struct Section: Equatable, Sendable {
    let id: CapabilitySectionID
    let items: [CapabilityItem]
    var label: String { id.label }
  }

  /// The typed action's half of the document. Every `AgentAction.Kind` must
  /// appear here exactly once; the test pins that.
  static func kindSection(_ kind: AgentAction.Kind) -> CapabilitySectionID {
    switch kind {
    case .showBalance, .showBudget, .openCardControls, .openReceive, .reply:
      return .instant
    case .askTransferDetails, .proposeTransfer, .proposeBudgetUpdate, .requirementsFlow, .agentTask:
      return .prepares
    case .transferReceipt:
      // The record that a commit landed — shown after the approval, never an
      // action of its own.
      return .approval
    }
  }

  static func kindLabel(_ kind: AgentAction.Kind) -> String {
    switch kind {
    case .showBalance: return "your balances"
    case .showBudget: return "this week's budget and the reserve"
    case .openCardControls: return "the card and its controls"
    case .openReceive: return "your receiving details"
    case .reply: return "a short explanation or a follow-up"
    case .askTransferDetails: return "a transfer, with the missing detail asked for"
    case .proposeTransfer: return "a transfer, ready for your confirmation"
    case .proposeBudgetUpdate: return "a change to the weekly budget"
    case .requirementsFlow: return "the requirements checklist when no provider is connected"
    case .agentTask: return "a research or comparison task"
    case .transferReceipt: return "the receipt after you approve a transfer"
    }
  }

  static var sections: [Section] {
    let kinds = AgentAction.Kind.allCases
    func kindItems(_ section: CapabilitySectionID) -> [CapabilityItem] {
      kinds
        .filter { kindSection($0) == section }
        .map { CapabilityItem(id: $0.rawValue, label: kindLabel($0)) }
    }
    func engineItems(_ section: CapabilityClass) -> [CapabilityItem] {
      MoneyDeskEngine.allCases
        .filter { $0.capabilityClass == section }
        .map { CapabilityItem(id: $0.rawValue, label: $0.label) }
    }

    return [
      Section(
        id: .instant,
        items: kindItems(.instant) + engineItems(.observe)),
      Section(
        id: .prepares,
        items: kindItems(.prepares) + engineItems(.prepare)),
      Section(
        id: .approval,
        items: CommitVerb.allCases.map { CapabilityItem(id: $0.rawValue, label: $0.label) }
          + kindItems(.approval)),
      Section(
        id: .research,
        items: ResearchEngine.allCases.map { CapabilityItem(id: $0.rawValue, label: $0.label) }),
      Section(
        id: .refuses,
        items: CapabilityRefusal.allCases.map { CapabilityItem(id: $0.rawValue, label: $0.label) }),
    ]
  }

  static func items(_ id: CapabilitySectionID) -> [CapabilityItem] {
    sections.first { $0.id == id }?.items ?? []
  }

  /// The answer, rendered from the document — never a second list to keep in
  /// step. The same five sections the proxy answers from.
  static var answer: String {
    func names(_ id: CapabilitySectionID) -> String {
      items(id).map(\.label).joined(separator: ", ")
    }
    return [
      "This build is simulated money, and it is honest about the line. Instantly, from your own records: \(names(.instant)).",
      "Prepared, with nothing moved: \(names(.prepares)).",
      "One approval, then it moves: \(names(.approval)).",
      "\(items(.research).count) engines run as research on the Mac runtime, and every one of them is preparation only — no vendor is connected.",
      "I refuse \(items(.refuses).map(\.label).joined(separator: ", ")).",
    ].joined(separator: " ")
  }

  /// The plain shapes of "tell me about this build". The proxy's typed read
  /// catches paraphrases; this answers the plain phrasings with no network.
  static func asksAboutThisBuild(_ message: String) -> Bool {
    let patterns = [
      #"\bwhat can (?:you|this|the app|this build|the product|it|mira)\b"#,
      #"\bwhat (?:does|do) (?:this|you|the app|mira|the build|it)(?:\s+\w+){0,2}\s+(?:do|offer|support|handle)\b"#,
      #"\bwhat (?:can't|can’t|cannot|don't|do not|doesn't|does not) (?:you|this|the app|mira|it|this build)\b"#,
      #"\bwhat (?:actually )?moves? money\b"#,
      #"\bmove money (?:in|here|in this build)\b"#,
      #"\bwhat(?:'s| is) simulated\b"#,
      #"\bwhat needs (?:my|your|one|an?) approval\b"#,
      #"\bwhat do you do\b"#,
      #"\bcapabilit(?:y|ies)\b"#,
      #"\bwhat are you (?:able|allowed) to do\b"#,
      #"\bhow does (?:mira|this|the app) work\b"#,
      #"\bwhat does the user see\b"#,
    ]
    return patterns.contains {
      message.range(of: $0, options: [.regularExpression, .caseInsensitive]) != nil
    }
  }
}
